### PR TITLE
add native view_copy.out ops, teach codegen about tensorlist out=

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR adds `out=` variants for all of the existing `view_copy` operators. I added them all manually in native_functions.yaml, and gave them implementations in `TensorShape.cpp`

Our codegen required some changes to handle them properly, because this is the first time we've had native functions that are `out=`, with the out parameter being a list of tensors. As Ed pointed out in this issue https://github.com/pytorch/pytorch/pull/76049, the return type of these functions should actually be `void`, and not `Tensor(a!)[]` - because you can't do method chaining on the output (since it's a list), then there's nothing useful about returning something.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #76358
* #77714
* #77713
* #77132
* #77716
* #77130
* #77129
* #77128
* #77126
* #77125
* #76320
* __->__ #76126

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -2777,5 +2777,264 @@ at::Tensor diagonal_scatter(const at::Tensor& self, const at::Tensor& src, int64
...
+
+
+at::Tensor& slice_copy_Tensor_out(const at::Tensor & self, int64_t dim, c10::optional<int64_t> start, c10::optional<int64_t> end, int64_t step, at::Tensor & out) {
+  auto tmp = self.slice(dim, start, end, step);
+  out.copy_(tmp);
+  return out;
+}
+
+
+at::ITensorListRef split_copy_Tensor_out(const at::Tensor & self, int64_t split_size, int64_t dim, at::TensorList  out) {
```
Comment (bdhirsh) [inline aten/src/ATen/native/TensorShape.cpp:2899]: @ezyang / @ysiraichi This PR isn't ready for review yet, but I'm wondering - what do you think the dispatcher return type should be for functions like this, with the following schema?
```
split_copy.Tensor_out(Tensor self, int split_size, int dim=0, *, Tensor(a!)[] out) -> Tensor(a!)[]
```

tldr is that this PR introduces some new ops that have a schema that the codegen hasn't had to deal with before - out= arguments that are lists of mutable tensors. I think the answer is that we want it to be `ITensorListRef`, but that currently breaks - so I wanted to clarify my reasoning.

Just to list out all the options:

(1) make the return type `ITensorListRef`

This breaks boxed_to_unboxed code right now, because we don't currently know how to convert `IValue` to `ITensorListRef`. I think... this is probably what we want? But it looks like this doesn't work today, so I was wondering what the current state of `ITensorListRef` is (or maybe I'm misunderstanding something - shouldn't you be able to convert an `IValue` to an `ITensorListRef`?)

(2) make the return type `std::vector<Tensor>`

This will work, but has two disadvantages: We have a bunch of codegen that expects the out= argument and the return type to be the same, so that we can just `return out` in the generated kernels. We'd have to modify those to explicitly convert the output to a vector (`return out.vec()`), and also accept the perf hit (I don't expect any of these kernels to be hotpath though).

(3) make the return type `at::TensorList`

This also breaks today, because `IValue`s don't know how to convert to a `TensorList` (`ArrayRef<Tensor>`). I think it would also be... wrong, because we'd need to convert from the IValue's internal `c10::List<Tensor>` to `ArrayRef<Tensor>`, which requires creating new tensors to make them all contiguous in memory, but they aren't owned by anyone.

Comment (ezyang) [reply aten/src/ATen/native/TensorShape.cpp:2899]: I was just discussing this at https://github.com/pytorch/pytorch/pull/76049 We should make the return type void. This would be BC breaking but I doubt anyone is actually using this API and in any case they can't do method chaining like this.

File: tools/codegen/api/cpp.py [hunk 2]
```diff
@@ -201,9 +202,12 @@ def returntype_type(t: Type, *, mutable: bool) -> CType:
         elif t.name == BaseTy.Scalar:
             return BaseCType(scalarT)
     elif isinstance(t, ListType):
-        elem = returntype_type(t.elem, mutable=mutable)
+        elem = returntype_type(t.elem, mutable=False)
```
Comment (ezyang) [inline tools/codegen/api/cpp.py:205]: maybe you want to assert that mutable is false here?

File: tools/codegen/api/cpp.py [hunk 3]
```diff
@@ -201,9 +202,12 @@ def returntype_type(t: Type, *, mutable: bool) -> CType:
...
         assert t.size is None, f"fixed size list returns not supported: {t}"
-        return VectorCType(elem)
+        if mutable and t.elem == BaseType(BaseTy.Tensor):
+            return BaseCType(iTensorListRefT)
```
Comment (ezyang) [inline tools/codegen/api/cpp.py:208]: it's weird to compute elem and then ignore it in this branch

**Comment by ezyang** [comment]
unsubbing till this is ready for review

**Comment by bdhirsh** [comment]
yep, sorry about that (everything in this stack except this PR + the LTC prototype are either ready to be landed soon or ready for review - you've reviewed most of them though)

File: tools/codegen/api/cpp.py (hunk 2 above)
Comment (bdhirsh) [reply tools/codegen/api/cpp.py]: yep, thanks! (this was a relic of me originally trying to return `Tensor(a!)[]` in the first version of the PR, but now these functions don't return anything).

**Comment by bdhirsh** [comment]
PR should be ready for review now (letting CI run again, but I think I've fixed tests + lints)

**Review by ezyang — APPROVED** [review]
With the assumption that these are getting gen'ed later.
